File: scripts/fetch-deals.mjs
// =========================================================
// 쿠팡파트너스 오픈API로 가격을 주기적으로 기록해두고,
// "예전 가격 대비 제일 많이 할인된 상품"이 맨 위에 오도록
// 자동으로 핫딜 목록을 만드는 스크립트.
//
// 실행: node scripts/fetch-deals.mjs
// 필요한 환경변수:
//   COUPANG_ACCESS_KEY  (파트너스 대시보드에서 발급받은 Access Key)
//   COUPANG_SECRET_KEY  (파트너스 대시보드에서 발급받은 Secret Key)
//
// GitHub Actions에서는 이 값들을 절대 코드에 직접 쓰지 말고
// Settings > Secrets and variables > Actions 에 등록해서 사용하세요.
//
// 동작 방식:
//  1. keywords.json에 있는 각 키워드로 상품을 검색
//  2. 검색된 모든 상품의 "현재 가격"을 price-history.json에 계속 쌓아둠
//     (쿠팡 검색API는 정가/할인율을 따로 안 줘서, 우리가 직접 시간별로
//      가격을 관찰해서 "예전보다 싸졌다"를 판단해야 해요)
//  3. 기록이 2번 이상 쌓인 상품 중에서, 예전 최고가 대비 지금 가격이
//     MIN_DISCOUNT 이상 떨어진 상품만 골라냄
//  4. 할인율이 높은 순서로 정렬 → 제일 많이 할인된 상품이 맨 위로
// =========================================================

import { createHmac } from "node:crypto";
import { readFile, writeFile } from "node:fs/promises";

const ACCESS_KEY = process.env.COUPANG_ACCESS_KEY;
const SECRET_KEY = process.env.COUPANG_SECRET_KEY;
const SUB_ID = process.env.COUPANG_SUB_ID || "hotdealsite"; // 파트너스 채널 구분용 (선택)

const DOMAIN = "https://api-gateway.coupang.com";
const SEARCH_PATH = "/v2/providers/affiliate_open_api/apis/openapi/products/search";

const SEARCH_LIMIT = 30;           // 키워드당 검색해서 가격을 기록해둘 상품 개수
const MIN_DISCOUNT = 0.35;         // 이 비율(35%) 이상 떨어진 상품만 핫딜 후보로 인정
const HISTORY_KEEP = 10;           // 상품별로 최근 몇 개의 가격 기록을 남길지
const DEAL_EXPIRY_HOURS = 48;      // 마지막 관측 후 이 시간이 지나면 목록에서 자동 제외
const REQUEST_DELAY_MS = 700;      // 요청 사이 대기 시간 (API 과호출 방지)

const KEYWORDS_FILE = new URL("../data/keywords.json", import.meta.url);
const HISTORY_FILE = new URL("../data/price-history.json", import.meta.url);
const OUTPUT_FILE = new URL("../data/deals-coupang.json", import.meta.url);

function assertEnv() {
  if (!ACCESS_KEY || !SECRET_KEY) {
    console.error(
      "❌ COUPANG_ACCESS_KEY / COUPANG_SECRET_KEY 환경변수가 없습니다.\n" +
      "   로컬 테스트라면: COUPANG_ACCESS_KEY=xxx COUPANG_SECRET_KEY=yyy node scripts/fetch-deals.mjs"
    );
    process.exit(1);
  }
}

// 쿠팡 공식 문서 기준 HMAC 서명 생성
function buildAuthHeader(method, pathWithQuery) {
  const [path, query = ""] = pathWithQuery.split("?");
  const now = new Date();
  const datetime =
    now
      .toISOString()
      .replace(/[-:]/g, "")
      .slice(2, 15) + "Z"; // yyMMddTHHmmssZ (UTC)

  const message = datetime + method + path + query;
  const signature = createHmac("sha256", SECRET_KEY)
    .update(message)
    .digest("hex");

  return `CEA algorithm=HmacSHA256, access-key=${ACCESS_KEY}, signed-date=${datetime}, signature=${signature}`;
}

async function searchProducts(keyword) {
  const query = `keyword=${encodeURIComponent(keyword)}&limit=${SEARCH_LIMIT}&subId=${encodeURIComponent(SUB_ID)}`;
  const pathWithQuery = `${SEARCH_PATH}?${query}`;
  const url = DOMAIN + pathWithQuery;

  const res = await fetch(url, {
    method: "GET",
    headers: {
      Authorization: buildAuthHeader("GET", pathWithQuery),
      "Content-Type": "application/json",
    },
  });

  if (!res.ok) {
    console.error(`⚠️ [${keyword}] 요청 실패: ${res.status} ${res.statusText}`);
    return [];
  }

  const json = await res.json();
  return json?.data?.productData || [];
}

async function loadJSON(fileUrl, fallback) {
  try {
    const text = await readFile(fileUrl, "utf-8");
    return JSON.parse(text);
  } catch {
    return fallback;
  }
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

async function main() {
  assertEnv();

  const { keywords = [] } = await loadJSON(KEYWORDS_FILE, { keywords: [] });
  const history = await loadJSON(HISTORY_FILE, {});
  const now = Date.now();

  console.log(`🔍 ${keywords.length}개 키워드 검색 시작...`);

  for (const keyword of keywords) {
    const products = await searchProducts(keyword);
    console.log(`  · "${keyword}" → ${products.length}건`);

    for (const p of products) {
      const id = String(p.productId);
      const entry = history[id] || {
        name: p.productName,
        image: p.productImage,
        url: p.productUrl,
        isRocket: !!p.isRocket,
        keyword,
        records: [],
      };

      entry.name = p.productName;
      entry.image = p.productImage;
      entry.url = p.productUrl;
      entry.isRocket = !!p.isRocket;
      entry.keyword = keyword;
      entry.records.push({ price: p.productPrice, at: now });
      entry.records = entry.records.slice(-HISTORY_KEEP);

      history[id] = entry;
    }

    await sleep(REQUEST_DELAY_MS);
  }

  await writeFile(HISTORY_FILE, JSON.stringify(history, null, 2), "utf-8");
  console.log(`💾 가격 기록 저장 완료 (${Object.keys(history).length}개 상품)`);

  // ---- 할인율 계산해서 핫딜 후보 뽑기 ----
  const candidates = [];

  for (const [id, entry] of Object.entries(history)) {
    const records = entry.records;
    if (records.length < 2) continue; // 비교할 예전 기록이 없으면 패스

    const latest = records[records.length - 1];
    const isStale = now - latest.at > DEAL_EXPIRY_HOURS * 3600 * 1000;
    if (isStale) continue; // 최근에 관측 안 된 상품은 제외

    const priorMax = Math.max(...records.slice(0, -1).map((r) => r.price));
    if (priorMax <= 0) continue;

    const discountRatio = (priorMax - latest.price) / priorMax;
    if (discountRatio >= MIN_DISCOUNT) {
      candidates.push({
        id,
        name: entry.name,
        price: latest.price,
        discount: Math.round(discountRatio * 100),
        source: "쿠팡",
        image: entry.image,
        link: entry.url,
        isRocket: entry.isRocket,
        keyword: entry.keyword,
        postedAt: new Date(latest.at).toISOString(),
      });
    }
  }

  // 할인율 높은 순으로 정렬 → 제일 많이 할인된 상품이 맨 위
  candidates.sort((a, b) => b.discount - a.discount);

  await writeFile(OUTPUT_FILE, JSON.stringify(candidates, null, 2), "utf-8");
  console.log(`🔥 할인 상품 ${candidates.length}건을 deals-coupang.json에 기록했습니다 (할인율 높은 순).`);
  if (candidates[0]) {
    console.log(`   1위: ${candidates[0].name} — ${candidates[0].discount}% ↓`);
  }
}

main().catch((err) => {
  console.error("스크립트 실행 중 오류:", err);
  process.exit(1);
});

File: fetch_deals.py
import os
import json
import time
import hmac
import hashlib
import requests
from datetime import datetime, timedelta, timezone

KST = timezone(timedelta(hours=9))
DATA_FILE = 'data.json'

def format_price(price):
    """가격 숫자/문자열에서 .0 소수점 제거 및 천단위 콤마 처리"""
    if price is None or price == "":
        return ""
    try:
        # float로 들어오거나 숫자 문자열일 경우 정수로 변환
        price_int = int(float(price))
        return f"{price_int:,}원"
    except (ValueError, TypeError):
        # 숫자로 변환 실패 시 .0원 문자열만 찾아 제거
        price_str = str(price).replace('.0원', '원').replace('.0', '')
        if not price_str.endswith('원') and price_str.isdigit():
            price_str += '원'
        return price_str

# ==========================================
# 1. 쿠팡 파트너스 수집 로직 (가격 .0원 제거 적용)
# ==========================================
def generate_coupang_signature(method, url, secret_key, access_key):
    path, _, query = url.partition('?')
    datetime_str = time.strftime('%y%m%dT%H%M%SZ', time.gmtime())
    message = datetime_str + method + path + (query if query else '')
    
    signature = hmac.new(
        bytes(secret_key, 'utf-8'),
        message.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()
    
    return f"CEA algorithm=HmacSHA256, access-key={access_key}, signed-date={datetime_str}, signature={signature}"

def fetch_coupang_goldbox():
    access_key = os.environ.get('COUPANG_ACCESS_KEY')
    secret_key = os.environ.get('COUPANG_SECRET_KEY')
    
    if not access_key or not secret_key:
        print("⚠️ [쿠팡] API Key가 설정되지 않았습니다.")
        return []

    domain = "https://api-gateway.coupang.com"
    url_path = "/v2/providers/affiliate_open_api/apis/openapi/v1/products/goldbox"
    
    authorization = generate_coupang_signature("GET", url_path, secret_key, access_key)
    headers = {
        "Authorization": authorization,
        "Content-Type": "application/json;charset=UTF-8"
    }

    try:
        response = requests.get(domain + url_path, headers=headers, timeout=10)
        if response.status_code == 200:
            result = response.json()
            products = result.get('data', [])
            
            coupang_items = []
            for item in products:
                orig_price_raw = item.get('originalPrice', 0)
                sale_price_raw = item.get('productPrice', 0)
                
                # 할인율 계산
                discount_rate = ""
                try:
                    orig_val = float(orig_price_raw) if orig_price_raw else 0
                    sale_val = float(sale_price_raw) if sale_price_raw else 0
                    if orig_val > sale_val > 0:
                        rate = round((1 - (sale_val / orig_val)) * 100)
                        discount_rate = f"{rate}%"
                except Exception:
                    pass

                coupang_items.append({
                    "id": f"coupang_{item.get('productId')}",
                    "source": "coupang",
                    "title": item.get('productName'),
                    "price": format_price(sale_price_raw), # .0원 제거 포맷팅 적용
                    "originalPrice": format_price(orig_price_raw) if orig_price_raw else "",
                    "discountRate": discount_rate,
                    "imageUrl": item.get('productImage'),
                    "link": item.get('productUrl'),
                    "createdAt": datetime.now(KST).isoformat()
                })
            print(f"✅ [쿠팡] 골드박스 상품 {len(coupang_items)}개 수집 완료!")
            return coupang_items
        else:
            print(f"❌ [쿠팡] API 호출 실패 (코드: {response.status_code})")
            return []
    except Exception as e:
        print(f"❌ [쿠팡] 수집 에러: {e}")
        return []

# ==========================================
# 2. 토스쇼핑 Open API 수집 로직
# ==========================================
BASE_TOSS_URL = "https://sharelink-api.toss.im"

def fetch_toss_deals():
    access_key = os.environ.get('TOSS_ACCESS_KEY')
    secret_key = os.environ.get('TOSS_SECRET_KEY')
    
    print("🔍 [토스] API 수집 시작...")
    if not access_key or not secret_key:
        print("⚠️ [토스] TOSS_ACCESS_KEY 또는 TOSS_SECRET_KEY 가 GitHub Secrets에 없습니다.")
        return []

    token = None
    try:
        auth_url = f"{BASE_TOSS_URL}/openapi/auth/token"
        res = requests.post(auth_url, json={"accessKey": access_key, "secretKey": secret_key}, timeout=10)
        if res.status_code == 200:
            res_json = res.json()
            data = res_json.get('data') or {}
            token = data.get('accessToken') if isinstance(data, dict) else res_json.get('accessToken')
            print("✅ [토스] Access Token 발급 성공!")
        else:
            print(f"❌ [토스] 토큰 발급 응답 실패 (코드: {res.status_code}) - {res.text}")
            return []
    except Exception as e:
        print(f"❌ [토스] 토큰 요청 예외: {e}")
        return []

    if not token:
        print("❌ [토스] Token 값이 비어있습니다.")
        return []

    headers = {"Authorization": f"Bearer {token}"}
    toss_items = []

    target_endpoints = [
        "/openapi/v1/products/daily-deals",
        "/openapi/products/daily-deals",
        "/openapi/v1/products",
        "/openapi/products"
    ]

    products = []
    for ep in target_endpoints:
        try:
            r = requests.get(f"{BASE_TOSS_URL}{ep}", headers=headers, timeout=10)
            if r.status_code == 200:
                body = r.json()
                items = body.get('data') or body.get('products') or []
                if isinstance(items, list) and len(items) > 0:
                    products = items
                    print(f"✅ [토스] 엔드포인트({ep})에서 {len(products)}개 상품 확보!")
                    break
        except Exception:
            continue

    if not products:
        print("⚠️ [토스] 오픈 API에서 유효한 상품 목록을 가져오지 못했습니다.")
        return []

    for item in products:
        prod_id = item.get('id') or item.get('productId')
        if not prod_id:
            continue

        short_url = None
        try:
            link_res = requests.post(
                f"{BASE_TOSS_URL}/openapi/v1/links",
                headers=headers,
                json={"productId": str(prod_id)},
                timeout=5
            )
            if link_res.status_code != 200:
                link_res = requests.post(
                    f"{BASE_TOSS_URL}/openapi/links",
                    headers=headers,
                    json={"productId": str(prod_id)},
                    timeout=5
                )

            if link_res.status_code == 200:
                l_body = link_res.json()
                l_data = l_body.get('data') or {}
                short_url = l_data.get('shortUrl') if isinstance(l_data, dict) else l_body.get('shortUrl')
        except Exception:
            pass

        final_link = short_url or item.get('linkUrl') or f"https://toss.im/shopping/product/{prod_id}"
        orig_price_raw = item.get('originalPrice', 0)
        sale_price_raw = item.get('price') or item.get('productPrice') or 0
        discount_rate = f"{item.get('discountRate')}%" if item.get('discountRate') else ""

        toss_items.append({
            "id": f"toss_{prod_id}",
            "source": "toss",
            "title": item.get('name') or item.get('title') or item.get('productName'),
            "price": format_price(sale_price_raw),
            "originalPrice": format_price(orig_price_raw) if orig_price_raw else "",
            "discountRate": discount_rate,
            "imageUrl": item.get('imageUrl') or item.get('thumbnail') or item.get('productImage'),
            "link": final_link,
            "createdAt": datetime.now(KST).isoformat()
        })

    print(f"🎉 [토스] 최종 {len(toss_items)}개 상품 처리 완료!")
    return toss_items

# ==========================================
# 3. 메인 실행 및 24시간 만료 관리
# ==========================================
if __name__ == '__main__':
    now = datetime.now(KST)
    
    try:
        with open(DATA_FILE, 'r', encoding='utf-8') as f:
            deals = json.load(f)
    except Exception:
        deals = []

    # 24시간 지난 만료 상품 삭제
    valid_deals = []
    for deal in deals:
        created_at_str = deal.get('createdAt')
        if created_at_str:
            try:
                created_at = datetime.fromisoformat(created_at_str)
                if now - created_at < timedelta(hours=24):
                    valid_deals.append(deal)
            except ValueError:
                continue

    # 신규 데이터 수집
    new_coupang = fetch_coupang_goldbox()
    new_toss = fetch_toss_deals()
    all_new = new_coupang + new_toss

    existing_links = {d['link'] for d in valid_deals}
    added_count = 0
    for item in all_new:
        if item['link'] not in existing_links:
            valid_deals.insert(0, item)
            added_count += 1

    with open(DATA_FILE, 'w', encoding='utf-8') as f:
        json.dump(valid_deals, f, ensure_ascii=False, indent=2)

    print(f"🎉 총 {len(valid_deals)}개 핫딜 저장 완료! (신규 추가: {added_count}개)")
